### PR TITLE
Overhaul GroovyScript

### PR DESCRIPTION
**Pull request by yor42**
## 📝 Description

This PR overhauls Grs support to fix and improve several things:
- Fix possible issues during script reloading
- Add link Generator and in-game wiki generator for better documentation

## 🎯 Goals

- Make current MWC Documented on Grs wiki
- Make Script writing easier with extra QOL methods (typing string for each enum is prone for human error and can be pain to debug)
- Make documentation easier to understand with in-game wiki generator feature

## ❌ Non Goals
- Change other part of mod's behavior

## 🚦 Testing 

Tested ``/gs generateWiki`` and ``/gs generateExample``, and generated example code does not throw errors.

## ⏮️ Backwards Compatibility 

Breaks existing Grs script as recipebuilder was updated to new Grs standard

## 📖 Added to documentation?

- [ ] 📜 README.md
- [x] 📑 Documentation
- [x] 📓 Javadoc
- [ ] 🍕 Comments
- [ ] 🙅 No documentation needed

### CODE REVIEW

File: src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java [hunk 1]
```diff
@@ -2,54 +2,60 @@
...
 
+@RegistryDescription(linkGenerator = ProjectConstants.ID)
 public class CraftingStation extends VirtualizedRegistry<IModernCraftingRecipe> {
 
     @Override
     public void onReload() {
-        this.removeScripted().forEach(CraftingRegistry::deleteRecipeRegistry);
-        this.restoreFromBackup().forEach(CraftingRegistry::registerRecipe);
+        this.removeScripted().forEach(CraftingRegistry::registerRecipe);
+        this.restoreFromBackup().forEach(CraftingRegistry::deleteRecipeRegistry);
```
Comment (Desoroxxx) [inline src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java]: This sounds wrong can you explain a bit what this does, like it sounds conflicting

**Review by Desoroxxx — CHANGES_REQUESTED** [review]

File: src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java [hunk 2]
```diff
@@ -97,48 +234,97 @@ public String getErrorMsg() {
...
-         * @param yield return Yield of upcoming ingredients
+         * @param ingredient crafting ingredient
+         * @param yield refund yield of ingredient
+         * @return AbstractRecipeBuilder
+         */
+        @RecipeBuilderMethodDescription(field = "input")
+        public AbstractRecipeBuilder<GSCrafting> input(double yield, IIngredient ingredient) {
+            if (yield < 0) {
+                throw new IllegalArgumentException("Yield cannot be negative, yet we got: " + yield);
+            }
```
Comment (brachy84) [inline src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java]: Dont throw. Either silently set to 0 or add warning during validation

File: src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java [hunk 3]
```diff
@@ -148,18 +334,12 @@ public RecipeBuilder setGroup(String group) {
...
             }
-
-            final ArrayList<CraftingEntry> entries = new ArrayList<>();
-
-            for (int i = 0; i < this.input.size(); i++) {
-                final IIngredient ingredient = this.input.get(i);
-                entries.add(new CraftingEntry(ingredient.toMcIngredient(), ingredient.getAmount(), this.yields.get(i)));
-            }
-
+            final ArrayList<CraftingEntry> entries = new ArrayList<>(this.input);
```
Comment (brachy84) [inline src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java]: The old way was fine (with the exception of proper builder methods)

File: src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java [hunk 4]
```diff
@@ -58,35 +64,166 @@ public void remove(IIngredient ingredient) {
...
+            @Example(".input(1, item('minecraft:clay')).output(item('minecraft:diamond'))"),
+            @Example(".input(0.5, item('minecraft:gold_ingot')).output(item('minecraft:clay') * 2)")
+    })
     public RecipeBuilder recipeBuilder() {
         return new RecipeBuilder();
     }
 
+    @Property(property = "group", comp = @Comp(not = "null"))
+    @Property(property = "input", comp = @Comp(gt = 0))
+    @Property(property = "output", comp = @Comp(eq = 1))
```
[the remark this answers is not in the document]
Comment (brachy84) [reply src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java:231]: Damn, ai actually giving decent feedback

File: src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java [hunk 5]
```diff
@@ -2,54 +2,60 @@
...
-    public void remove(IIngredient ingredient) {
-        for (ArrayList<IModernCraftingRecipe> list : craftingMap.values()) {
-            for (IModernCraftingRecipe recipe : list) {
-                if (ingredient.test(recipe.getItemStack())) {
-                    deleteRecipeRegistry(recipe);
-                    this.addBackup(recipe);
-                }
-            }
+    @MethodDescription(example = @Example("ore('oreDiamond')"), priority = 2000)
+    public void removeAllCategoryByOutput(IIngredient ingredient) {
```
Comment (brachy84) [inline src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java]: the name is confusing
we are not removing categories
`removeByOuput` is fine

File: src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java (hunk 1 above)
Comment (brachy84) [reply src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java]: Yeah, this looks wrong. You are supposed to remove previously via script added recipes. And add back those which were removed

**Review by brachy84 — CHANGES_REQUESTED** [review]

**Review by Desoroxxx — CHANGES_REQUESTED** [review]
LGTM, still a comment left unresolved

**Review by Desoroxxx — APPROVED** [review]

**Review by Desoroxxx — APPROVED** [review]

**Pull request merged**